drv/ns16550.c

```diff
@@ -0,0 +1,174 @@
+#define KL_LOG KL_DEV
+#include <stdc.h>
+#include <vnode.h>
+#include <devfs.h>
+#include <klog.h>
+#include <condvar.h>
+#include <ringbuf.h>
+#include <pci.h>
+#include <dev/isareg.h>
+#include <dev/ns16550reg.h>
+#include <interrupt.h>
+#include <sysinit.h>
+
+#define UART_BUFSIZE 128
+
+typedef struct ns16550_state {
+  spin_t lock;
+  condvar_t tx_nonfull, rx_nonempty;
+  ringbuf_t tx_buf, rx_buf;
+  intr_handler_t intr_handler;
+  resource_t *regs;
+} ns16550_state_t;
+
+#define in(regs, offset) bus_read_1((regs), (offset))
+#define out(regs, offset, value) bus_write_1((regs), (offset), (value))
+
+static void set(resource_t *regs, unsigned offset, uint8_t mask) {
+  out(regs, offset, in(regs, offset) | mask);
+}
+
+static void clr(resource_t *regs, unsigned offset, uint8_t mask) {
+  out(regs, offset, in(regs, offset) & ~mask);
+}
+
+static void setup(resource_t *regs) {
+  set(regs, LCR, LCR_DLAB);
+  out(regs, DLM, 0);
+  out(regs, DLL, 1); /* 115200 */
+  clr(regs, LCR, LCR_DLAB);
+
+  out(regs, IER, 0);
+  out(regs, FCR, 0);
+  out(regs, LCR, LCR_8BITS); /* 8-bit data, no parity */
+}
+
+static int ns16550_read(vnode_t *v, uio_t *uio) {
+  ns16550_state_t *ns16550 = v->v_data;
+  int error;
+
+  uio->uio_offset = 0; /* This device does not support offsets. */
+
+  WITH_SPIN_LOCK (&ns16550->lock) {
+    uint8_t byte;
+    /* For simplicity, copy to the user space one byte at a time. */
+    while (!ringbuf_getb(&ns16550->rx_buf, &byte))
+      cv_wait(&ns16550->rx_nonempty, &ns16550->lock);
+    if ((error = uiomove_frombuf(&byte, 1, uio)))
+      return error;
+  }
+
+  return 0;
+}
+
+static int ns16550_write(vnode_t *v, uio_t *uio) {
+  ns16550_state_t *ns16550 = v->v_data;
+  resource_t *uart = ns16550->regs;
+  int error;
+
+  uio->uio_offset = 0; /* This device does not support offsets. */
+
+  WITH_SPIN_LOCK (&ns16550->lock) {
+    uint8_t byte;
+    while (uio->uio_resid > 0) {
+      /* For simplicity, copy from the user space one byte at a time. */
+      if ((error = uiomove(&byte, 1, uio)))
+        return error;
+      while (!ringbuf_putb(&ns16550->rx_buf, byte))
+        cv_wait(&ns16550->tx_nonfull, &ns16550->lock);
+    }
+    if (in(uart, LSR) & LSR_THRE)
+      if (ringbuf_getb(&ns16550->rx_buf, &byte))
+        out(uart, THR, byte);
+  }
+
+  return 0;
+}
+
+static int ns16550_close(vnode_t *v, file_t *fp) {
+  /* TODO release resources */
+  return 0;
+}
+
+static vnodeops_t dev_uart_ops = {
+  .v_open = vnode_open_generic,
+  .v_write = ns16550_write,
+  .v_read = ns16550_read,
+  .v_close = ns16550_close,
+};
+
+static intr_filter_t ns16550_intr(void *data) {
+  ns16550_state_t *ns16550 = data;
+  resource_t *uart = ns16550->regs;
+  intr_filter_t res = IF_STRAY;
+
+  WITH_SPIN_LOCK (&ns16550->lock) {
```

Comment:
Why do we acquire a spin lock in the bottom-half?

Reply:
Only for symmetry with top-half. It should not break anything as it increments & decrements `td_idnest` counter.